File: staticfiles/js/subscription/manage_subscription.js
document.addEventListener('DOMContentLoaded', function() {
    const cancelSubscriptionBtn = document.getElementById('cancelSubscriptionBtn');
    const confirmCancelSubscriptionBtn = document.getElementById('confirmCancelSubscriptionBtn');
    const cancellationReason = document.getElementById('cancellationReason');
    const otherReasonContainer = document.getElementById('otherReasonContainer');
    const subscriptionStatusBadge = document.getElementById('subscriptionStatusBadge');

    // Функция для получения текущей даты в формате DD.MM.YYYY
    function getCurrentDate() {
        const now = new Date();
        const day = String(now.getDate()).padStart(2, '0');
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const year = now.getFullYear();
        return `${day}.${month}.${year}`;
    }

    // Функция для показа уведомлений
    function showAlert(message, type) {
        // Здесь можно использовать любую библиотеку для уведомлений, например, Bootstrap Toast
        alert(message); // Временная заглушка
    }

    // Обработчик для кнопки отмены/возобновления подписки
    if (cancelSubscriptionBtn) {
        cancelSubscriptionBtn.addEventListener('click', function() {
            const isCancelled = cancelSubscriptionBtn.classList.contains('btn-outline-success');
            if (!isCancelled) {
                // Показываем модальное окно для отмены
                const modal = new bootstrap.Modal(document.getElementById('cancelSubscriptionModal'));
                modal.show();
            } else {
                // Возобновляем подписку
                fetch('/manage-subscription/', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'X-CSRFToken': document.querySelector('[name=csrfmiddlewaretoken]').value
                    },
                    body: JSON.stringify({ action: 'renew' })
                })
                .then(response => response.json())
                .then(data => {
                    if (data.status === 'success') {
                        subscriptionStatusBadge.classList.remove('bg-warning');
                        subscriptionStatusBadge.classList.add(data.badge_class);
                        subscriptionStatusBadge.textContent = data.new_status;
                        cancelSubscriptionBtn.classList.remove('btn-outline-success');
                        cancelSubscriptionBtn.classList.add(data.button_class);
                        cancelSubscriptionBtn.innerHTML = `<i class="ph ph-x me-2"></i> ${data.button_text}`;
                        showAlert(data.message, 'success');
                    } else {
                        showAlert(data.message, 'danger');
                    }
                })
                .catch(error => {
                    console.error(window.translations.error ||'Ошибка:', error);
                    showAlert(window.translations.error_message || 'Произошла ошибка при возобновлении подписки', 'danger');
                });
            }
        });
    }

    // Обработчик для подтверждения отмены подписки
    if (confirmCancelSubscriptionBtn) {
        confirmCancelSubscriptionBtn.addEventListener('click', function() {
            const reasonNumber = cancellationReason.value;
            if (!reasonNumber) {
                showAlert(window.translations.reason_required || 'Пожалуйста, выберите причину отмены', 'danger');
                return;
            }
            let reasonText = '';

            if (reasonNumber === '1') {
                reasonText = window.translations.too_expensive;
            } else if (reasonNumber === '2') {
                reasonText = window.translations.not_use_service;
            } else if (reasonNumber === '3') {
                reasonText = window.translations.missing_features;
            } else if (reasonNumber === '4') {
                reasonText = window.translations.found_alternative;
            } else if (reasonNumber === '5') {
                reasonText = document.getElementById('otherReason').value || window.translations.other;
            }

            // Закрываем модальное окно
            const modal = bootstrap.Modal.getInstance(document.getElementById('cancelSubscriptionModal'));
            modal.hide();

            // Отправляем запрос на сервер
            fetch('/manage-subscription/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': document.querySelector('[name=csrfmiddlewaretoken]').value
                },
                body: JSON.stringify({
                    action: 'cancel',
                    reason_number: reasonNumber,
                    other_reason: reasonNumber === '5' ? reasonText : '',
                    cancellation_date: getCurrentDate()
                })
            })
            .then(response => response.json())
            .then(data => {
                if (data.status === 'success') {
                    subscriptionStatusBadge.classList.remove('bg-success');
                    subscriptionStatusBadge.classList.add(data.badge_class);
                    subscriptionStatusBadge.textContent = data.new_status;
                    cancelSubscriptionBtn.classList.remove('btn-outline-danger');
                    cancelSubscriptionBtn.classList.add(data.button_class);
                    cancelSubscriptionBtn.innerHTML = `<i class="ph ph-arrows-clockwise me-2"></i> ${data.button_text}`;
                    showAlert(data.message, 'info');
                } else {
                    showAlert(data.message, 'danger');
                }
            })
            .catch(error => {
                console.error(window.translations.error ||'Ошибка:', error);
                showAlert(window.translations.error_message || 'Произошла ошибка при отмене подписки', 'danger');
            });
        });
    }

    // Показ поля для другой причины отмены
    if (cancellationReason && otherReasonContainer) {
        cancellationReason.addEventListener('change', function() {
            otherReasonContainer.style.display = this.value === '5' ? 'block' : 'none';
        });
    }
});

File: apps/subscribe/manage_subscription.py
from django.views.decorators.http import require_POST
from project.language import translate, language
from django.http import JsonResponse
from django.utils import timezone
import json


# Маппинг номеров причин для сервера
REASON_MAPPING = {
    '1': 'too_expensive',
    '2': 'not_using',
    '3': 'missing_features',
    '4': 'found_alternative',
    '5': 'other'
}

@require_POST
def manage_subscription(request):
    """Обработка отмены или возобновления подписки"""
    global subscription_status
    try:
        data = json.loads(request.body)
        action = data.get('action')
        reason_number = data.get('reason_number', '')  # Получаем номер причины
        cancellation_date = data.get('cancellation_date', timezone.now().strftime('%d.%m.%Y'))

        if action == 'cancel':
            if not reason_number or reason_number not in REASON_MAPPING:
                return JsonResponse({
                    'status': 'error',
                    'message': translate('Пожалуйста, выберите причину отмены', request.LANGUAGE_CODE)
                }, status=400)
            subscription_status = False
            reason_text = REASON_MAPPING.get(reason_number, 'other')
            other_reason = data.get('other_reason', '') if reason_number == '5' else ''
            # Логируем с учетом текста для причины "Другое"
            log_message = f"Подписка отменена. Причина: {reason_number}"
            if reason_number == '5' and other_reason:
                log_message += f" ({other_reason})"
            log_message += f". Дата отмены: {cancellation_date}"
            print(log_message)
            return JsonResponse({
                'status': 'success',
                'message': translate('Подписка будет отменена', request.LANGUAGE_CODE) + f". {translate('Причина', request.LANGUAGE_CODE)}: {reason_text if reason_number != '5' else other_reason}",
                'new_status': f"{translate('Будет отменена', request.LANGUAGE_CODE)} {cancellation_date}",
                'button_text': translate('Возобновить подписку', request.LANGUAGE_CODE),
                'button_class': 'btn-outline-success',
                'badge_class': 'bg-warning'
            })
        elif action == 'renew':
            subscription_status = True
            print(f"Подписка возобновлена. Время: {timezone.now().strftime('%d.%m.%Y %H:%M:%S')}")
            return JsonResponse({
                'status': 'success',
                'message': translate('Подписка успешно возобновлена', request.LANGUAGE_CODE),
                'new_status': translate('Активна', request.LANGUAGE_CODE),
                'button_text': translate('Отменить подписку', request.LANGUAGE_CODE),
                'button_class': 'btn-outline-danger',
                'badge_class': 'bg-success'
            })
        else:
            return JsonResponse({'status': 'error', 'message': 'Недопустимое действие'}, status=400)
    except Exception as e:
        print(f"Ошибка в manage_subscription: {str(e)}")
        return JsonResponse({'status': 'error', 'message': str(e)}, status=500)
